### ISSUE TITLE
Ways to contribute to glTF

### ISSUE THREAD

**Issue by pjcozzi** [KhronosGroup/glTF#456]
We believe the true usefulness of glTF goes beyond the spec itself; it is an ecosystem of tools, documentation, and extensions contributed by the community.

We encourage everyone to contribute to the glTF community.  We are happy to list your work in this repo's main [README.md](https://github.com/KhronosGroup/glTF/blob/master/README.md).

Way to contribute include:

**Content Pipeline Tools**

* Export glTF from Blender, Maya, etc.
* Convert popular formats to glTF.
* Optimize glTF assets, e.g., texture atlas, batching, etc.
* Add glTF support to [Assimp](http://assimp.sourceforge.net/).

**Runtime**

* Simple minimal loader for WebGL, OpenGL ES, or OpenGL.
* Helper libraries for loading glTF, e.g., decompression extension.

**Documentation**

* Getting started with glTF.
* Tutorial on your experience implementing glTF.
* Size/performance of glTF compared to other formats.

**Extensions**

* Implement and propose [extensions](https://github.com/KhronosGroup/glTF/blob/master/extensions/README.md) to the core spec.

**Models**

* Contribute [sample models](https://github.com/KhronosGroup/glTF/tree/master/sampleModels) for testing.

If you want us to spread the word about your glTF work, leave a comment here or email Patrick Cozzi, pjcozzi@siggraph.org.

**Comment by tparisi** [comment]
How are people going to find this document?

**Comment by pjcozzi** [comment]
It is linked from the main README.md and from https://www.khronos.org/gltf

**Comment by tparisi** [comment]
Excellent



On Mon, Oct 19, 2015 at 9:37 AM, Patrick Cozzi <notifications@github.com>
wrote:

> It is linked from the main README.md and from https://www.khronos.org/gltf
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/KhronosGroup/glTF/issues/456#issuecomment-149274724>.
>



-- 


Tony Parisi                             tparisi@gmail.com
Follow me on Twitter!             http://twitter.com/auradeluxe
Read my blog at                     http://www.tonyparisi.com/
Learn WebGL                         http://learningwebgl.com/
Mobile                                    415.902.8002
Skype                                     auradeluxe

Read my books!
*Learning Virtual Reality*
http://www.amazon.com/Learning-Virtual-Reality-Experiences-Applications/dp/1491922834


*Programming 3D Applications in HTML5 and
WebGLhttp://www.amazon.com/Programming-Applications-HTML5-WebGL-Visualization/dp/1449362966
<http://www.amazon.com/Programming-Applications-HTML5-WebGL-Visualization/dp/1449362966>WebGL,
Up and Running*
http://www.amazon.com/dp/144932357X

**Comment by vinjn** [comment]
I have a WIP project assimp2gltf
https://github.com/vinjn/assimp2gltf

The purpose of this project: convert files in 40+ 3D file formats, including Collada, 3DS, OBJ, LWO, FBX, Blender, X, STL, PLY, MS3D, B3D, MD3, MDL, DXF and IFC to gltf.

However it's far far away from that purpose.

**Comment by pjcozzi** [comment]
@vinjn wow, this has great potential.  I see that the last commit was 6 months ago.  Do you plan to update this now that glTF has reached draft 1.0?  Here is the glTF 0.8 to 1.0 guide: https://github.com/KhronosGroup/glTF/wiki/glTF-0.8-to-1.0-Guide

**Comment by phaazon** [comment]
I work on a Haskell loader, [here](https://github.com/phaazon/gltf). I was surprised to see that the format uses numeric constant instead of labels (GL_FLOAT instead of 9349364, for instance).

**Comment by pjcozzi** [comment]
@phaazon sounds cool!  Let us know when it is ready and we'll list it [here](https://github.com/KhronosGroup/glTF/blob/master/README.md#gltf-tools).

glTF uses constants instead of strings since they are the easiest thing for the runtime to use.  See #96.

**Comment by phaazon** [comment]
Yeah, I see. I’ll try to advance on my loader and will keep you know – I have a lot of project in the kitchen, but I will eventually have the loader ready sooner or later.

**Comment by vinjn** [comment]
@pjcozzi Yes, I have the plan. And I want to use glTF as the scene description format for my hobby engine project. I will inform you of the progress.

**Comment by games** [comment]
A year ago I use glTF(based on 0.6) in my project(https://github.com/games/orange), seems I should upgrade to 1.0 :)

**Comment by pjcozzi** [comment]
@games yes, please update and let us know when your engine supports 1.0!

**Comment by kimkulling** [comment]
Hi,

AssetImporterLib will create a new importer for you. See https://github.com/assimp/assimp/issues/673

**Comment by pjcozzi** [comment]
@kimkulling great news!  This is exciting.  Let me know if you have any questions on the spec or need sample code.

**Comment by pjcozzi** [comment]
@kimkulling do you plan to write a glTF exporter for assimp as well?  That would be really useful for the WebGL community.

**Comment by erlend-sh** [comment]
I suggest you add to "Content Pipeline Tools":

- Create glTF output for [fbx-conv](https://github.com/libgdx/fbx-conv)

**Comment by pjcozzi** [comment]
@erlend-sh added!  Thanks for bringing this project to our attention.

**Comment by kimkulling** [comment]
Exporter would be a great thing, but let start with the importer first :-).

**Comment by kimkulling** [comment]
Thanks to Otger Rogla we have a first experimental glTF-Loader in Asset-Importer-Lib!

**Comment by pjcozzi** [comment]
Wow, @otgerp @kimkulling this is great news!  Congrats and thanks for sharing!

When should we list it with the [glTF Tools](https://github.com/KhronosGroup/glTF/blob/master/README.md) in this repo, e.g., does it need doc, waiting on an official release, etc?

**Comment by vinjn** [comment]
Thanks @otgerp @kimkulling! That's super fast.

**Comment by vinjn** [comment]
A new C++ loader
https://github.com/syoyo/tinygltfloader

**Comment by pjcozzi** [comment]
> A new C++ loader https://github.com/syoyo/tinygltfloader

Thanks for sharing!  This is great, @syoyo!

**Comment by syoyo** [comment]
Cool!

**Comment by pjcozzi** [comment]
@kimkulling @otgerp I just came across https://github.com/assimp/assimp/pull/699!  So assimp can export glTF now too, right?  Wow!  I will update the glTF README.md.

**Comment by kimkulling** [comment]
Yes we can ☺. Thanks for the read me update.

Patrick Cozzi <notifications@github.com> schrieb am Di., 5. Jan. 2016 15:13:

**Comment by vinjn** [comment]
https://github.com/tparisi/Unity-glTF-Exporter from @tparisi

**Comment by pjcozzi** [comment]
@tparisi is the Unity exporter ready for us to list in this repo's README.md?

**Comment by pjcozzi** [comment]
Hi everyone following this issue - we are preparing a glTF community update for GDC.  Please let me know if you have glTF work that we can highlight.  Thanks!

**Comment by erlend-sh** [comment]
Just found out there's now a Blender exporter in the works by @Moguri

https://github.com/Kupoman/blendergltf

Found it because apparently glTF support is one of the suggested projects for [Blender GSoC 2016](http://wiki.blender.org/index.php/Dev:Ref/GoogleSummerOfCode/2016/Ideas).

**Comment by pjcozzi** [comment]
Amazing news!  Really looking forward to this, @Moguri, @ideasman42 and @Kupoman.

**Comment by erlend-sh** [comment]
@cyrillef just made FBX-glTF compatible with the 1.0 spec as of this commit: https://github.com/cyrillef/FBX-glTF/pull/3

**Comment by 5chdn** [comment]
I am writing a toolset of geographic web applications utilizing gltf, however, some questions remain open. If anyone could look over unanswered questions on stack exchange?

**Comment by pjcozzi** [comment]
@5chdn looks like all those questions have answers now.  Are there any that need more info?

**Comment by AVGP** [comment]
Just FYI: We have added glTF import and export to Archilogic: https://spaces.archilogic.com/blog/gltf-import-export

**Comment by feiss** [comment]
Small tool for batch converting COLLADA to glTF using collada2gltf: https://github.com/feiss/batchgltf

**Comment by pjcozzi** [comment]
Thanks @feiss!  Added to #867.  All of these tools will make there way to the main README.md in the next big update.

**Comment by AVGP** [comment]
Here's a web component to display glTF models on the web like this:

```html
<gltf-viewer src="model/duck.gltf" interactive></gltf-viewer>
```

See the [repository](https://github.com/avgp/gltf-viewer) or [live demo](https://avgp.github.io/gltf-viewer) for more.

**Comment by pjcozzi** [comment]
Very nice @AVGP, thanks for sharing!  Could you please open a pull request to add it to https://github.com/KhronosGroup/glTF#loaders-and-viewers?

**Comment by jratcliff63367** [comment]
So, I've been Google searching all over and I cannot find any project to implement a GLTF importer for UE4.  Is anyone aware of an existing project that does this?  I could, of course, write one from scratch, but that seems foolish if someone else is already actively working on such a project.

Also, my use case is less about using GLTF to import graphics assets but more for physics assets.  Does GLTF already have a fully fledged physics asset data definition? If so, who has been most involved in that?  If it does not, what do I need to do to get involved in helping create that specification?  I am a Principal Engineer at NVIDIA Corporation who works extensively with physics content and I want to evaluate GLTF as a suitable candidate for transporting physics asset data.

My alternative approach is to create custom FBX nodes, but that seems far more hacky than simply leveraging against GLTF which is kind of a 'fresh start' at this point.

**Comment by donmccurdy** [comment]
> Does GLTF already have a fully fledged physics asset data definition?

@jratcliff63367 No official glTF physics extension yet, but some folks in the Blender and AltspaceVR communities have been working on a vendor extension: https://github.com/Kupoman/blendergltf/tree/master/extensions/BLENDER_physics

**Comment by sbtron** [comment]
As per this tweet - https://twitter.com/C4gIo/status/907267375392624641 
This project is being developed as a glTF importer for UE4 https://github.com/code4game/libgltf

**Comment by pjcozzi** [comment]
@jratcliff63367 any contributions you make to the glTF community are much appreciated!

Please email me - pjcozzi@gmail.com - if you would like to join any of the weekly Khronos glTF calls to discuss physics, UE4, or anything else.

**Comment by pjcozzi** [comment]
If anyone is looking for a fun and hopefully quick glTF project, @jratcliff63367 suggested writing a glTF import/export extension for the [FBX Extensions SDK](http://help.autodesk.com/view/FBX/2017/ENU/?guid=__files_GUID_E03E1C4B_FAD9_4E7B_9BBB_4F6782282402_htm) to facilitate glTF tapping into the FBX ecosystem.

Let us know if you work on this!

**Comment by ousttrue** [comment]
I am writing a unity project.
I am focusing on humanoid model.

https://github.com/ousttrue/UniGLTF

![humanoid](https://user-images.githubusercontent.com/68057/35196012-abc21cd2-ff0f-11e7-9404-1d2c37eea416.gif)

**Comment by pjcozzi** [comment]
Cool, thanks for sharing @ousttrue!  I added a note to https://github.com/KhronosGroup/glTF/issues/1058#issuecomment-359596961.  Keep us posted on your progress and feel free to connect with @sbtron and crew who are working on https://github.com/KhronosGroup/UnityGLTF if there are any opportunities for collaboration.

**Comment by sbtron** [comment]
Great work @ousttrue. Have you taken a look at https://github.com/KhronosGroup/UnityGLTF. We are still looking at implementing animation support. Let us know if you are interested in integrating/collaborating further.

**Comment by donmccurdy** [comment]
Not sure what would be the right place to put tools that don't exist but I wish were available, I'd love to have a **glTF-Optimizer** glTF equivalent of https://github.com/jonnenauha/obj-simplify. I use that tool for merging geometry and otherwise optimizing models and, for lack of importers  into DCC tools, it's currently much harder to do that with glTF. The tool could be created separately and then bundled with https://github.com/AnalyticalGraphicsInc/gltf-pipeline, perhaps.

**Comment by nickkraakman** [comment]
Hi, we're working on integrating glTF conversion into a microservice framework, so we built upon the amazing [ClayGL fbx2gltf converter](https://github.com/pissang/claygl/blob/master/tools/fbx2gltf.py) (which also support OBJ and COLLADA inputs), and wrapped it in a web API. 

The API can be tested live here at https://gltfapi.co

The source code is published on GitHub here: https://github.com/headjack-dev/gltf-api

**Comment by pjcozzi** [comment]
@nickkraakman very cool and thanks for sharing, I made a note in this issue: https://github.com/KhronosGroup/glTF/issues/1058#issuecomment-370006615

**Comment by funfoo** [comment]
Hi! I am the founder of 3D Slash http://www.3dslash.net . 3D Slash is an easy to use 3d modeling software with its unique interface, as fun as a building game.

We've just release the glTF exporter, we hope that this will allow more users to express themselves in 3D, and maybe to share on FB.

Could you add us in the Applications section of the README.md? We export 2.0, and the mention could be "Web-based and app-based 3d modeling imitating the stone-cutter"

All the best,
Sylvain Huet

**Comment by huxingyi** [comment]
Hi, for anyone who may be interested, especially game developers, I have added glTF2.0 exporter for my tiny opensource quick 3D modeling tool https://github.com/huxingyi/dust3d, the exported content inside .gltf file include the auto-generated meshes, skeleton, and materials.

**Comment by gildor2** [comment]
Hi all,

I have a project (UModel - https://github.com/gildor2/UModel) which allows to view and export models with animations from Unreal engine games. Recently it got a possibility to export data in glTF 2.0 format. No idea if it is suitable to be mentioned for glTF community, however I think, in the worst case glTF importer / plugin developers can use it to test how their systems will work with complex AAA quality game models which have lots of animations.

Thanks,
Konstantin

**Comment by pjcozzi** [comment]
@gildor2 that is fantastic, congrats!  Can you please open a pull request to add UModel to the list of exporters: https://github.com/KhronosGroup/glTF/blob/master/README.md#gltf-tools?

**Comment by kestrelm** [comment]
Hello all,

The Creature 2D Skeletal + Mesh Animation Tool now suppots GLTF export.
Check out this tweet for the details:
https://twitter.com/KestrelmMoon/status/1040671251742638081

It showcases exported GLTF Character animation running in BabylonJS and Godot Engine 3.x . Should work with any tool that supports GLTF skinned mesh character animation.

Cheers

**Comment by jeyashankher** [comment]
Hello All,

We are building a cloud based to convert 3D assets to glTF at Point105-AR (https://www.point105ar.com). Our focus is to create a feature-set of cloud-based features and optimizations for 3D artists to easily create their 3D content.

A few features we now have on our service:

1) Support for specialised materials:
  - V-Ray for 3DS Max support (we have a machine learnt algorithm to convert from V-Ray workflow to Metallic Roughness)
  - Maya StingRay
  - 3DS Max Physical Material, currently it is similar to BabylonJS export. We do plan to move to the native Physical Material workflow.

2) We support a few automatic integrations for texture creation and optimization
- Auto generate ORM texture
- Bump to Normal maps, Auto generated transparent maps
- Automatic image compression

We currently take in as input a single embedded binary FBX file as input to generate glTF.

**Comment by vpenades** [comment]
Hi all,

I'm working on a glTF library for C#, it's been in development for a while, and there's still a lot of work to do... but I feel confident enough to show it to the world.

For anyone interested, the repository is [here](https://github.com/vpenades/SharpGLTF).

**Comment by emackey** [comment]
@vpenades Looks cool, can you open a PR to add this to README.md?  Looks like this would get added to the [list of C# projects](https://github.com/KhronosGroup/glTF#c-1).

Anyone else on this thread is also welcome to open PRs for glTF-related stuff too, thanks!

**Comment by vpenades** [comment]
@emackey I've just updated the library to a new version, I think I am confident now to include it in the list of c# projects

**Comment by emackey** [comment]
Added in #1594

**Comment by patriciogonzalezvivo** [comment]
Hi all! I add loading of gltf2.0 files to [GlslViewer](https://github.com/patriciogonzalezvivo/glslViewer) through tinyGltf and I'm loving it! 
Here are some videos loading and rendering in different RaspberryPi models https://twitter.com/patriciogv/status/1181264063096971264?s=20

**Comment by khyveasna** [comment]
`Jan giving me a chance to get`

**Comment by mikikg** [comment]
Hi all, here is my contribution, **Android** port of **jgltfViewer** including ported **jglTF** library with native OpenGL ES 3.0 bindings:
https://github.com/mikikg/AndroidJgltfViewer
Work in progress, please contribute.

**Comment by trejkaz** [comment]
I had some general feedback about the format as one who wanted to implement a parser.

I found that glTF with separate files was quite nice and intuitive.

But for GLB I found that having to deal with both a text format and a binary format in one file was cumbersome. For parsing JSON chunks we do have nice libraries. For parsing binary we also have nice libraries such as Kaitai Struct. But when you're writing a definition of something in Kaitai Struct, you want to do things like, "this binary element is repeated N times where N is stored over here", and if "over here" is inside a giant blob of JSON, now you can't do it entirely in a declarative binary format anymore and have to break out to specific programming languages to parse the JSON and make it available to the binary parsing side.

A format which is 100% binary and well structured like EBML would really have been nicer. In fact, EBML already exists, is extensible, and works great for all sorts of binary, so I think it would have been a decent choice for a base format, rather than inventing yet another new container.

**Comment by Lilothestitch16** [comment]
Hi. Will there be a GLTF Importer to 3ds Max?

**Comment by javagl** [comment]
@Lilothestitch16 This repository and this issuse are the wrong place (and the wrong place (sic)) to ask this. websearches reveal things like https://forums.autodesk.com/t5/3ds-max-forum/gltf-import/m-p/12253731/highlight/true#M209638 , and in general, forums (or other 3DS Max support channels) are a better place for questions like this.
